cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -818,13 +847,80 @@ private Location copyProgramJar() throws IOException {
     return programJarCopy;
   }
 
+  /**
+   * Creates a JAR file that contains all the plugin jars that are needed by the job. The plugin directory
+   * structure is maintained inside the jar so that MR framework can correctly expand and recreate the required
+   * structure.
+   *
+   * @return the {@link Location} for the archive file or {@code null} if there is no plugin files need to be localized
+   */
+  @Nullable
+  private Location createPluginArchive(@Nullable AdapterSpecification adapterSpec,
+                                       Id.Program programId) throws IOException {
+    if (adapterSpec == null) {
+      return null;
+    }
+
+    Set<PluginInfo> pluginInfos = adapterSpec.getPluginInfos();
+    if (pluginInfos.isEmpty()) {
+      return null;
+    }
+
+    File templateDir = new File(cConf.get(Constants.AppFabric.APP_TEMPLATE_DIR));
+    File pluginDir = new File(templateDir, adapterSpec.getTemplate());
+    File jarFile = File.createTempFile("plugin", ".jar");
+    try {
+      JarOutputStream output = new JarOutputStream(new FileOutputStream(jarFile));
+      try {
+        // Create the directory entries
+        output.putNextEntry(new JarEntry(adapterSpec.getTemplate() + "/"));
```

Comment:
Just checked that ``JarEntry`` doesn't take a ``File`` or something like that as input :(

Will this break on windows?